### PR TITLE
Correctly set formatters and locators on removed shared axis

### LINKED ISSUE #14911: Removing a shared axes via `ax.remove()` leads to an error.
### Bug report

**Bug summary**

removing a shared axes via `ax.remove()` leads to an error.

This bisects to "Fix locator/formatter setting when removing shared Axes" #13983 (in current master, targeted for 3.2) which initially headed out to make the situation of removing a shared axes better, fixing https://github.com/matplotlib/matplotlib/issues/12853, but apparently the test doesn't capture the below simple case.

Hence marking as release critical for 3.2, in order not to release this bug.

**Code for reproduction**

```python
import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, sharex=True)

axs[0].remove()

plt.show()
```

**Actual outcome**

```
Traceback (most recent call last):
  File "d:\***\matplotlib\lib\matplotlib\backends\backend_qt5.py", line 488, in _draw_idle
    self.draw()
  File "d:\***\matplotlib\lib\matplotlib\backends\backend_agg.py", line 396, in draw
    self.figure.draw(self.renderer)
  File "d:\***\matplotlib\lib\matplotlib\artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "d:\***\matplotlib\lib\matplotlib\figure.py", line 1722, in draw
    renderer, self, artists, self.suppressComposite)
  File "d:\***\matplotlib\lib\matplotlib\image.py", line 136, in _draw_list_compositing_images
    a.draw(renderer)
  File "d:\***\matplotlib\lib\matplotlib\artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "d:\***\matplotlib\lib\matplotlib\axes\_base.py", line 2622, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "d:\***\matplotlib\lib\matplotlib\image.py", line 136, in _draw_list_compositing_images
    a.draw(renderer)
  File "d:\***\matplotlib\lib\matplotlib\artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "d:\***\matplotlib\lib\matplotlib\axis.py", line 1227, in draw
    ticks_to_draw = self._update_ticks()
  File "d:\***\matplotlib\lib\matplotlib\axis.py", line 1103, in _update_ticks
    major_locs = self.get_majorticklocs()
  File "d:\***\matplotlib\lib\matplotlib\axis.py", line 1348, in get_majorticklocs
    return self.major.locator()
  File "d:\***\matplotlib\lib\matplotlib\ticker.py", line 2032, in __call__
    return self.tick_values(vmin, vmax)
  File "d:\***\matplotlib\lib\matplotlib\ticker.py", line 2040, in tick_values
    locs = self._raw_ticks(vmin, vmax)
  File "d:\***\matplotlib\lib\matplotlib\ticker.py", line 1979, in _raw_ticks
    nbins = np.clip(self.axis.get_tick_space(),
  File "d:\***\matplotlib\lib\matplotlib\axis.py", line 2193, in get_tick_space
    length = ((ends[1][0] - ends[0][0]) / self.axes.figure.dpi) * 72
AttributeError: 'NoneType' object has no attribute 'dpi'
```

**Expected outcome**

A plot with the first axes removed, as would be achieved with the same code running in 3.1

**Matplotlib version**

  * Operating system: Win 8.1
  * Matplotlib version: current master
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.6

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #14911 

The issue was that a shared axis might never have it's `Formatter/Locator` set, which is fine when it is shared, but when the axis storing the `Formatter/Locator` it relies on is removed, the previous shared axis needs to have the `Formatter/Locator`s set.

Also fixed typo `_reset_axis_form` > `_reset_axis_from`.

I've tried a few things, but cannot come up with a more elegant solution than this; if anyone has any ideas suggestions welcome.

### CODE REVIEW

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -1596,25 +1596,37 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             return axarr
 
     def _remove_ax(self, ax):
-        def _reset_loc_form(axis):
+        def _reset_loc_from(axis):
```
Comment (jklymak) [inline lib/matplotlib/figure.py]: Why did you change the name?

Comment (dstansby) [reply lib/matplotlib/figure.py]: Because the function is resetting locators from another axis, hence the shortening `reset_loc_from`.

**Comment by jklymak** [comment]
Ok so I think this is a case of sharex=True meaning too many things.  Are we garaunteeibg that sharing the axis means we are sharing the locators and formatters?  That seems really hard to enforce and when you toss units into the mess, it becomes almost intractable.  I agree that just removing an axis shouldn’t cause a crash, but I think the fundamental issue is that the locators and formatters were linked in the first place.

File: lib/matplotlib/figure.py (hunk 1 above)
Comment (jklymak) [reply lib/matplotlib/figure.py]: I think it’s reset locators formatters

**Comment by anntzer** [comment]
> Are we garaunteeibg that sharing the axis means we are sharing the locators and formatters?

At least IIRC @efiring has a fairly strong opinion that we do...

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -1596,25 +1596,37 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
...
+            # Because set_major_formatter() etc. force isDefault_* to be False,
+            # we have to manually check if the original formatter was a
+            # default and manually set isDefault_* if that was the case.
             majfmt = axis.get_major_formatter()
-            if not majfmt.axis.isDefault_majfmt:
-                axis.set_major_formatter(majfmt)
+            isDefault = majfmt.axis.isDefault_majfmt
+            axis.set_major_formatter(majfmt)
+            if isDefault:
+                majfmt.axis.isDefault_majfmt = True
```
Comment (jklymak) [inline lib/matplotlib/figure.py]: This doesn't make any sense to me.  You get the formatter for `axis` and then you set the formatter for `axis` to that formatter, and then you set the isDefault to the value axis had before.  I don't see how this code is doing anything.

Comment (dstansby) [reply lib/matplotlib/figure.py]: I think the confusion here stems from the mappings of formatters to axis.
- Axis > Formatter is a one to one mapping
- Formatter > Axis is a one to many mapping

So a single `Formatter` is linked to a single `Axis`, but many other `Axis` can make use of that `Formatter`. I agree this is a nightmare to understand, but without re-doing the internals of how this works it is how it is.

So this code is getting the formatter that `axis` is using, (that it might be borrowing) and then explicitly associating that formatter with `axis` (so it is no longer borrowing it).

**Comment by dstansby** [comment]
I appreciate this is messy code, but it is neccesitated by the weird way that `Formatter`/`Locator` and `Axis` interact. I think the options are
1. Merge this PR as a fix (+/- small changes)
2. Change the way `Formatter` and `Locator` work with shared `Axis`

I think 2. is a bit much to do for `3.2`, but maybe should be looked at for `3.3`?

File: lib/matplotlib/figure.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/figure.py]: should better be `_reset_locators_and_formatters`.

**Comment by jklymak** [comment]
Be good to rebase to double check that docs build, but I don't think that has anything to do w/ this PR.

File: lib/matplotlib/tests/test_figure.py [hunk 3]
```diff
@@ -480,3 +480,10 @@ def test_axes_removal():
     axs[0].plot([datetime(2000, 1, 1), datetime(2000, 2, 1)], [0, 1])
     assert isinstance(axs[0].xaxis.get_major_formatter(),
                       ScalarFormatter)
+
+
+def test_removed_axis():
+    # Simple smoke test to make sure removing a shared axis works
+    fig, axs = plt.subplots(2, sharex=True)
+    axs[0].remove()
+    fig.canvas.draw()
```
Comment (efiring) [inline lib/matplotlib/tests/test_figure.py]: Are you sure this doesn't work before this PR?  When I try, with `ipython --pylab` and mpl 3.1.1:
```
fig, axs = plt.subplots(2, sharex=True)
axs[0].remove()
```
nothing unexpected happens; the figure is redrawn with only the lower Axes.

Comment (efiring) [reply lib/matplotlib/tests/test_figure.py]: Ignore the above.  I see that the problem was introduced after 3.1.1.
